### DISCUSSION TITLE
GrainTracker issue

### DISCUSSION THREAD

**Discussion by ykvishal** [idaholab/moose#17572] [Q&A Modules: Phase field] [answered]
I am performing large 2D (2400X2400 mesh and 20K grains) and 3D (192X192X192, 10K grains) simulations. I am using  GrainTracker block from [grain_growth_2D_graintracker.i](https://github.com/idaholab/moose/blob/6437ec83fe011ae67c7b71b4210b426db6ca320b/modules/phase_field/examples/grain_growth/grain_growth_2D_graintracker.i)

>   [./grain_tracker]
>     type = GrainTracker
>     threshold = 0.2
>     connecting_threshold = 0.08
>     compute_halo_maps = true # Only necessary for displaying HALOS
>   [../]

When I was analyzing microstructure evolution, I noticed that two grains represented by  a same OP are merging (Grain coalescence) into one. A typical example is here:

![GrainTracker](https://user-images.githubusercontent.com/31565781/114321562-af282c80-9b1b-11eb-8c8a-32c838790d8f.jpg)

**This is happening for many grains.**

Earlier I had similar issue and I observed this issue even for small system size and small number of grains (600 grains). One animation is shown here:
https://user-images.githubusercontent.com/31565781/114320915-95391a80-9b18-11eb-90ab-4cfc3e363778.mp4
At that time, I tried few options like 1) Increase mesh size; 2) Play with  threshold  and connecting_threshold  parameters; 3) increase OPs. But these steps were not helpful. When I used a)  coloring_algorithm = bt and int_width=8 in the PolycrystalVoronoi block then it helped a lot.  Unfortunately, for large scale simulations, these two parameters have other effects like a) increase memory requirement and b) the 'projecting initial condition' step during initialization takes several hours (with 560 cores). It becomes worse for 60K grains in 2D and 20K grains in 3D. So I changed back these parameters (coloring_algorithm = jp and int_width=0 - it resulted in 7.5 times improvement in speed for projecting initial condition). Later, I tried to use Adaptivity  option but that was not useful in this case. Also adaptivity option takes several hours (with 560 cores) for large number of grains. 

If you have some suggestions or tips then please let me know.

Thank you.

**Comment by SudiptaBiswas** [comment, marked as answer]
Sorry, what is the issue here? Why shouldn't the grains coalesce? Do you have remapping set to true? What does the halos map look like? Maybe create a small problem to reproduce/explain the issue.

**Reply by ykvishal** [reply]
Thank you for your feedback. My response is as follows:

`Sorry, what is the issue here? Why shouldn't the grains coalesce?`

In my research, I am only considering curvature driven grain growth. In ideal grain growth, boundary migration occurs only due to curvature of the boundary. Growth of grains by  grains coalesce step is unphysical in an ideal grain growth. 

`Do you have remapping set to true?`

Yes. Default option is true. I did not touch the original GrainTracker block given above.

`What does the halos map look like? `

Here is the halos map
![GrainTrackerHalos](https://user-images.githubusercontent.com/31565781/114609271-e7a44380-9c9e-11eb-880b-d3d6b1db75d5.jpg)


`Maybe create a small problem to reproduce/explain the issue.`

My research involves large number of grains (>100K). I started with small system size and I managed to avoid grain coalescence for smaller cases. However, for large simulations, the parameter selected by me (as explained above) takes longer time to reach to t=0 step. So, I went back to default options to save time. I am not sure if solution for small size will work for large system.

**Reply by SudiptaBiswas** [reply]
It looks to me that the two grains are coalescing because of the shrinkage of the smaller grains that were previously present between these two grains. This often happens in polycrystals. However, if this is only happening because the two grains are represented by the same order parameters, then it could be a numerical issue. Probably, the remapping is not working properly. You might consider changing the threshold values or using a different seed for the initial condition.

**Reply by ykvishal** [reply]
`However, if this is only happening because the two grains are represented by the same order parameters, then it could be a numerical issue. `

Yes. That is the case. I should have mentioned that grains represented by same OP are coalescing very often. 

`You might consider changing the threshold values or using a different seed for the initial condition.`

I will try these suggestions. It seems that the parameters related to GrainTraker block are very sensitive to initial number of grains.

**Reply by ykvishal** [reply]
After some trials , now it is looking better. 

![Effect of GrainTracker settings](https://user-images.githubusercontent.com/31565781/114654909-7343c180-9ceb-11eb-8fcb-1c48bc8985f6.jpg)

It seems that current default settings may work with small number of grains only. 

Thank you for the discussion @SudiptaBiswas !!

**Comment by PengWei97** [comment]
Hi @ykvishal, I also encountered the problem of grain merging, and the effect is as follows,
![image](https://user-images.githubusercontent.com/56183775/177318062-3faa5f67-ed4d-4811-b806-a3245b6013da.png)
However, I didn't see the solution in your discussion, so could you tell me how to solve this problem?

wei

**Reply by ykvishal** [reply]
In my case, I varied halo level and checked this problem. However, the GrainTracker parameters are dependent on various factors and my solution may or may not work for you. You will have to spend a lot of time to optimize GrainTracker parameters.  All the best.

**Reply by PengWei97** [reply]
Thanks for your reply, I will try some of the settings mentioned in your discussion.

**Reply by mdh23rct** [reply]
@PengWei97 Dear Peng, I am having identical grain coalescence problem. Would you be kind to let me know the parameters you used in the GrainTracker userobject to avoid this grain coalescence?? Thanks.
